### ISSUE TITLE
The `--repl-no-load` option is not recognized for build-type: Custom

### ISSUE THREAD

**Issue by enobayram** [haskell/cabal#8379]
**Describe the bug**
I'm not able to start `cabal repl` with the `--repl-no-load` in Haskell packages specified to have `build-type: Custom`.

**To Reproduce**

I've prepared a GitHub repository to demonstrate this issue. So check out: https://github.com/enobayram/cabal-repl-no-load-test

Then in that folder, the following fails, even though a simple `cabal v2-repl` works.

```
$ cabal v2-repl --repl-no-load -v
...
creating
/***/cabal-no-load-test/dist-newstyle/build/x86_64-linux/ghc-8.10.7/cabal-no-load-test-0.1.0.0/setup
/***/.ghcup/bin/ghc --make -fbuilding-cabal-package -odir /***/cabal-no-load-test/dist-newstyle/build/x86_64-linux/ghc-8.10.7/cabal-no-load-test-0.1.0.0/setup -hidir /***/cabal-no-load-test/dist-newstyle/build/x86_64-linux/ghc-8.10.7/cabal-no-load-test-0.1.0.0/setup -i -i/***/cabal-no-load-test/. -optP-include -optP/***/cabal-no-load-test/dist-newstyle/build/x86_64-linux/ghc-8.10.7/cabal-no-load-test-0.1.0.0/setup/setup_macros.h -hide-all-packages -no-user-package-db -package-db /***/store/ghc-8.10.7/package.db -package-db /***/cabal-no-load-test/dist-newstyle/packagedb/ghc-8.10.7 -package-id Cabal-3.2.1.0 -package-id base-4.14.3.0 /***/cabal-no-load-test/dist-newstyle/build/x86_64-linux/ghc-8.10.7/cabal-no-load-test-0.1.0.0/setup/setup.hs -o /***/cabal-no-load-test/dist-newstyle/build/x86_64-linux/ghc-8.10.7/cabal-no-load-test-0.1.0.0/setup/setup -threaded
/***/cabal-no-load-test/dist-newstyle/build/x86_64-linux/ghc-8.10.7/cabal-no-load-test-0.1.0.0/setup/setup
repl --verbose=2
--builddir=/***/cabal-no-load-test/dist-newstyle/build/x86_64-linux/ghc-8.10.7/cabal-no-load-test-0.1.0.0
--repl-no-load exe:cabal-no-load-test
unrecognized 'repl' option `--repl-no-load'
CallStack (from HasCallStack):
  withMetadata, called at src/Distribution/Simple/Utils.hs:370:14 in Cabal-3.8.1.0-inplace:Distribution.Simple.Utils
Error: cabal: repl failed for cabal-no-load-test-0.1.0.0-inplace.
```

**Expected behavior**
`cabal v2-repl --repl-no-load` should drop me down to a GHCi session that has no modules loaded.

**System information**
 - Ubuntu 22.04
 - `cabal --version`: 3.8.1.0
 - `ghc --version`: 8.10.7

**Additional context**
Note that `cabal v2-repl --repl-no-load` works, if you change `build-type` to `Simple` in [cabal-no-load-test.cabal](https://github.com/enobayram/cabal-repl-no-load-test/blob/ed5550029f22d5e06be21ababb5063a3b73901f1/cabal-no-load-test.cabal#L42)

**Comment by ulysses4ever** [comment]
Hey @enobayram ! I’m really sorry it took so long to get back to you!

Thanks a lot for bundling up a reproducer, that will help a lot.

I believe currently, there’s very little energy for working on anything `build-type: Custom`-related. Some consider it a miss feature that eventually should be deprecated and removed. So, I hate to say this, but I’m not sure how much support you’ll get here. If you can look closer into `-v3` output and try to figure what’s going wrong proactively and ask questions, chances are you’d get some feedback.

**Comment by enobayram** [comment]
@ulysses4ever Thanks a lot for the response! Your input about `build-type: Custom` is valuable for me, leading me to believe that we should seek the solution in moving away from it.

**Comment by ulysses4ever** [comment]
@enobayram yes, avoiding it, if possible, is the preferred way to go forward.

**Comment by mpickering** [comment]
I imagine this issue is "fixed" now because the problem was probably because the Cabal version used to build the Setup didn't allow --repl-no-load, but we didn't force a new version of Cabal library when trying to use this feature. There is a similar constraints for enable-multi-repl which forces a new enough Cabal version.

**Comment by enobayram** [comment]
@mpickering Thanks a lot for circling back to this issue. I think you're right about your assessment, my repro project above starts a "no-load" REPL without any issues when I fix the `base` version bounds and run `cabal repl --repl-no-load` in it with `Cabal-3.10.3.0` + `ghc-9.10.1`.

**Comment by fendor** [comment]
Sounds like we can close this as fixed :)
